quinn-proto/src/connection/mod.rs

```diff
@@ -1004,6 +1043,29 @@ impl Connection {
                             .push_back(EndpointEventInner::NeedIdentifiers(now, num_new_cid));
                     }
                 }
+                Timer::MaxAckDelay => {
+                    trace!("max ack delay reached");
+                    // This timer is only armed in the Data space
+                    self.spaces[SpaceId::Data]
+                        .pending_acks
+                        .on_max_ack_delay_timeout()
+                }
+                Timer::Rtt => {
+                    // This timer is only armed when ACK frequency has been enabled, the peer
+                    // supports it, and we are in the Data space
+                    let space = &mut self.spaces[SpaceId::Data];
+                    trace!(
+                        in_flight_ack_eliciting = space.sent_packets.len(),
+                        "rtt timer"
+                    );
+
+                    if !space.sent_packets.is_empty() {
+                        // It is recommended to request an immediate ACK once per RTT if there are
+                        // unacked packets
+                        space.immediate_ack_pending = true;
+                    }
+                    self.timers.set(Timer::Rtt, now + self.path.rtt.get());
```

Comment:
Somehow I can't reply inline to [this comment](https://github.com/quinn-rs/quinn/pull/1553#discussion_r1249004850), so I'll do it here, as both are related.

First of all, I'm glad you found the root cause of the infinite loop. Sorry I missed that. It'd been better if I'd persevered looking for the problem, instead of tweaking the testing code.

Regarding a solution, the simplest one I can come up with is to schedule the next timeout after `max(RTT, TIMER_GRANULARITY)`. That way we can sidestep additional complexity, and to me it seems like a sensible choice This code is executed right after the RTT timeout has elapsed and it doesn't make sense to schedule the new RTT timeout before the next tick, which I assume happens after `TIMER_GRANULARITY`. Do you agree?

Otherwise, setting the timeout when sending an ack-eliciting packet would also work, but requires some bookkeeping (i.e. tracking the `Instant` of the last ACK sent, so we can check whether an RTT has elapsed since).

Reply:
> I can't reply inline

The input box for that can be found in the original thread. Threading interacts kind of weirdly with review-scoped comments.

> schedule the next timeout after max(RTT, TIMER_GRANULARITY).

I'm not sure how this solves the problem of the infinite loop, or of spurious wakeups when there's no activity. In the case of the looping test, RTT is 10ms, so replacing `now + rtt` with `now + max(rtt, TIMER_GRANULARITY)` won't change behavior at all. Did you mean to propose `now + rtt + TIMER_GRANULARITY`? That seems like an improvement, but latency variance we'll still likely see some superfluous transmits, which is suboptimal.

> setting the timeout when sending an ack-eliciting packet [...] requires some bookkeeping

Can't we just set the timer to `now + rtt` if it's unset, and leave it alone if it's already set? Did you mean to say tracking the `Instant` of the last ACK *received*, so we don't request redundant ACKs from the peer? That sounds like an improvement, but is independent of the two separate issues identified, which are:
1. IMMEDIATE_ACKs count as in-flight for purposes of sending another IMMEDIATE_ACK, causing spurious IMMEDIATE_ACK sends if the RTT timer expiry is handled before the peer's ACK is processed, which is likely to happen as the peer's ACK will take about an RTT to arrive.
2. The RTT timer is always active, even when the connection is idle, which causes unnecessary wakeups.

Reply:
> Sorry I missed that.

And no worries there, sometimes an outside perspective makes it a lot easier to diagnose stuff :grinning:

Reply:
Thanks for summarizing the problem again. It's helps load back part of the context in my head, after 2 months of working on other stuff 😅. Reading back what I wrote above, I'm not that sure it makes much sense. Now let me try to articulate my thoughts about both problems you mentioned.

### RTT is always active

The current implementation is too naive, waking the connection every RTT even if there is no work to do. Your suggested alternative sounds compelling. Here's how it would work in practice:

1. By default the timer should be inactive
2. When sending an ack-eliciting packet and the RTT timer is not set, we should set the timer to `now + RTT` _and_ request an immediate ack (otherwise there is a chance we won't get an ACK within RTT, and the draft says that we should _cause a receiver to send an acknowledgement at least once per RTT if there are unacknowledged ack-eliciting packets in flight_)
3. When the timer is elapsed _and_ there are ack-eliciting packets in-flight, we should set the timer to `now + RTT` _and_ request an immediate ack

Does that sound reasonable?

### Timeout-triggered immediate ack frames can cause an infinite loop

I'll reply inline in the thread 😉

Reply:
> Does that sound reasonable?

Sounds perfect!